include/deal.II/base/tensor_polynomials_base.h

```diff
@@ -0,0 +1,181 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2005 - 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_tensor_polynomials_base_h
+#define dealii_tensor_polynomials_base_h
+
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/exceptions.h>
+#include <deal.II/base/point.h>
+#include <deal.II/base/tensor.h>
+
+#include <vector>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * This class provides a framework for the finite element polynomial
+ * classes for use with finite element classes that are derived from
+ * @p FE_Poly and @p FE_PolyTensor. An object of this type is stored
+ * as a member variable in the finite element class.
+ *
+ * <h3>Deriving classes</h3>
+ *
+ * Any derived class must provide the most basic properties for shape
+ * functions evaluated on the reference cell. This includes, but is not
+ * limited to, implementing the <tt>compute</tt>, <tt>name</tt>, and
+ * <tt>clone</tt> member functions. These functions are necessary to store the
+ * most basic information of how the polynomials in the derived class evaluate
+ * at a given point on the reference cell. More information on each function can
+ * be found in the corresponding function's documentation.
+ *
+ * Some classes that derive from this class include
+ * <ul>
+ *   <li> <tt>PolynomialsABF</tt>
+ *   <li> <tt>PolynomialsBDM</tt>
+ *   <li> <tt>PolynomialsBernardiRaugel</tt>
+ *   <li> <tt>PolynomialsNedelec</tt>
+ *   <li> <tt>PolynomialsRaviartThomas</tt>
+ *   <li> <tt>PolynomialsRT_Bubbles</tt>
+ * </ul>
+ *
+ * @ingroup Polynomials
+ * @author Graham Harper
+ * @date 2019
+ */
+template <int dim>
+class TensorPolynomialsBase
+{
+public:
+  /**
+   * Constructor. This takes a integer @p k from the finite element class,
+   * which assigned to @p my_degree on construction. @p n is assigned to @p n_pols
+   * on construction.
```

Comment:
Drop the references to the names of member variables -- these are internal to the class, and not of interest to the reader.

Reply:
Also, can you use better names for `n` and `k`? You might want to say what they *represent*, rather than what variables they are assigned to.

Reply:
How about `deg` and `n`? I know `n` isn't super helpful, but I'm unsure what else would be a fitting name.

Reply:
We generally spell things out. How about `max_polynomial_degree` and `n_polynomials`?

Reply:
`n_polynomials` sounds good to me!